packages/dds/tree/src/simple-tree/treeNodeApi.ts

```diff
@@ -101,6 +104,51 @@ export interface TreeNodeApi {
 		listener: TreeChangeEvents[K],
 	): () => void;
 
+	/**
+	 * Register an event listener on an array node. This is a specialized version of {@link on} for array nodes, with
+	 * listener payload types specific to them.
+	 * @param node - The node whose events should be subscribed to.
+	 * @param eventName - Which event to subscribe to.
+	 * @param listener - The callback to trigger for the event. The tree can be read during the callback, but it is invalid to modify the tree during this callback.
+	 * @returns A callback function which will deregister the event.
+	 * This callback should be called only once.
+	 */
+	on<K extends keyof ArrayNodeChangeEvents>(
+		node: TreeArrayNode,
+		eventName: K,
+		listener: ArrayNodeChangeEvents[K],
+	): () => void;
+
+	/**
+	 * Register an event listener on an object node. This is a specialized version of {@link on} for object nodes, with
+	 * listener payload types specific to them.
+	 * @param node - The node whose events should be subscribed to.
+	 * @param eventName - Which event to subscribe to.
+	 * @param listener - The callback to trigger for the event. The tree can be read during the callback, but it is invalid to modify the tree during this callback.
+	 * @returns A callback function which will deregister the event.
+	 * This callback should be called only once.
+	 */
+	on<K extends keyof ObjectNodeChangeEvents>(
+		node: TreeObjectNode,
```

Comment:
TreeObjectNode requires generic type parameters but I'm having a hard time figuring out what should I pass.

Reply:
If you don't know anything about what fields a node has, all you can say about an object node is thats its API is the same as TreeNode, but it might vane some more fields. There is no way to capture that in the type system that doesn't also allow all non-object nodes. There is nothing that can be used to restrict an API to only accepting object nodes, or to derive if a node is an object node from its type.

You can make an API that requires you to provide an object node schema, and a matching node since object node schema can be distinguished by their NodeKind. Thats the best you could do here, and I assume thats not at all what you want.


If you want it to be possible to write an APi that accepts any object node, we would have to add some property thats unique to object nodes. For example we could add a NodeKind symbol to tree node, an statically type it as NodeKind.Object for Object nodes. Without a change like that to the public TreeNode or object node APS, I'm pretty sure waht you want is impossible.